designs/air-gapped.md

```diff
@@ -0,0 +1,97 @@
+1. Provision an OCI registry.
+1. Run `eksctl anywhere download artifacts` to download manifests to disk.
+1. Run `eksctl anywhere download images` to download images and charts to disk.
+1. Run `eksctl anywhere import images` to push container images and charts to a registry.
+1. Configure their cluster spec file to point to the registry 
+1. Run all cluster commands (`create`, `upgrade`, etc.) with the `--bundles-override` flag pointing to the `Bundles` manifest downloaded in the first step.
+
+## Overview of Solution
+1. Users will provide an OCI registry.
+1. We will store all dependencies (container images, helm charts and yaml manifests) in that registry.
+1. If a cluster config with a [`registryMirrorConfiguration`](https://github.com/aws/eks-anywhere/blob/main/pkg/api/v1alpha1/cluster_types.go#L51) is provided, the CLI and the rest of components will always pull them from it.
+No extra flags/configuration will be needed and the user will interact with this cluster as with any other one.
+
+## Solution Details
+
+### Populating the registry
+
+In order to populate the registry, we will offer two commands:
+```sh
+eksctl anywhere export artifacts --output artifacts.tar
+eksctl anywhere import artifacts --input artifacts.tar --registry myregistry.com
+```
+
+The first one downloads all dependencies (yaml manifests, images and helm charts). The only command argument here is the destination file. This command will create a tarball containing all 3 types of artifacts in that location.
+The second one unpackages the tarball created in the first command, reads the packaged `Bundles` manifest and imports the referenced dependencies to a registry. The input here will be the artifacts tarball, the registry endpoint and the registry credentials (provided through env vars).
+
+This two commands can be expanded in the future to add more capabilities like selectively exporting dependencies (eg. only for one kubernetes version or for only one provider).
+
+### Pre packaged dependencies
+
+An specific version of the CLI tied to one `Bundles` manifest will always produce the exact same artifacts tarball.
+This means that we can prepackage the dependencies, store them in a public bucket and reference the tarball in the `Release` manifest.
+This simplifies the experience for users who are interested in the default dependencies bundle.
+
+## Implementation
+
+The proposed design can be implemented in 4 incremental phases:
+1. Add the two new commands but only push images and charts to the registry. This will require to keep using the `--bundles-override` pointing to a `Bundles` manifest in disk. This is an incremental improvement over the current state of `download images` so work should be minimal.
+2. Push manifests to OCI registry and add the capability to the CLI to download them.
+3. Start storing our default manifests in public ECR as opposed to Cloudfront + S3. This unifies even more the behavior of connected and disconnected environments.
```

Comment:
Using a "yaml" mediatype (either reusing `application/vnd.oci.yaml-sample.config.v3+yaml` or just creating our own). I haven't looked deeply into the technical details, but I know it's possible to use an OCI registry to store "random" files.

I don't think it's orthogonal though. At least not to simplifying the UX. By doing this, users don't need to worry anymore about having to pass the `--bundles-override` flag and about having to have all yaml manifests stored in local disk.

If what you are suggesting is that we could possibly use a different storage mechanism and achieve the same UI, yeah that's true. But that would involve having two dependencies instead of one. Unless you have a better idea which I would be happy to talk about.

Reply:
I've seen it done before. I'm more saying just because we can, doesn't mean we should. OCI is about image formats, storing something that doesn't contain an image feels misleading. 

From the users point of view, they interact with 2 endpoints with respect to an image and a bundle. Is that really a problem?

Reply:
I don't agree with the fact that OCI is only about images formats, at least not anymore.
 
It might have started with that, but now a days we see other types of assets being stored in registries (mostly related with packaged software). Helm charts is one example.

There is even a [repo](https://github.com/opencontainers/artifacts) in the open containers org about this. You can even see SBOM mentioned there, which in a way, is _similar_ to our `Bundle`.

I could be in the wrong here but I don't think this is a "we do it because we can". At least in my opinion, it is the right solution.

As far as 2 endpoint vs 1 being a problem: that's tricky question. What is really a problem? That depends on the definition of problem, which is driven by previous experiences, expectations and a bit of personal preference.

Answering with mine, yes I think it is a problem. It's not about providing two flags, it's about managing two dependencies. That's double the work and double the things that can go wrong. We already got instances where managing a private registry was "too much effort".

Do you have any other concern (aside from being misleading) about using the registry? Technical or other.

Also, what alternative would you propose instead?

Reply:
> I don't agree with the fact that OCI is only about images formats, at least not anymore.

You're right, its expanded way beyond what I remember having combed through a bunch of resources.

How wide-spread is the support for custom media types? Is there an existing media type we can leverage or would we need to define our own?

Reply:
Scratch that "existing types" question. I see in an earlier reply some commentary.